DSMRloggerAPI.ino

```diff
@@ -373,119 +373,141 @@ void setup()
   }                                                                               //USE_NTP
-#endif                                                                            //USE_NTP
+  #endif                                                                            //USE_NTP
+}
 
+//===[ Do the background tasks ]===
+void doBackgroundTasks()
+{
+  httpServer.handleClient();
+  MDNS.update();
+  handleKeyInput();
+  handleMQTT();                 // MQTT transmissions
+  handleMindergas();            // Mindergas update 
 #if defined( HAS_OLED_SSD1306 ) || defined( HAS_OLED_SH1106 )
   checkFlashButton();
-  if (millis() - lastOledStatus > 5000) 
-  {
-    lastOledStatus = millis();
-    displayStatus();
-  }
-#endif
-
-  if (!showRaw) 
-  {
-    slimmeMeter.loop();
-    //---- capture new telegram ??
-    if (millis() > telegramInterval) 
-    {
-      telegramInterval = millis() + (settingInterval * 1000);  // test 10 seconden
-      slimmeMeter.enable(true);
-#ifdef ARDUINO_ESP8266_GENERIC
-      digitalWrite(LED_BUILTIN, LED_ON);
-#else
-      digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
-#endif
-    }
-  } // !showRaw 
-  
-  //---- this part is processed in 'normal' operation mode!
-  if (!showRaw) 
-  {
-  #if defined(HAS_NO_SLIMMEMETER)
-    handleTestdata();
-  #else
-    handleSlimmeMeter();
-  #endif
-  }
-  else   
-  {
-#if defined( HAS_OLED_SSD1306 ) || defined( HAS_OLED_SH1106 )
-      if (showRawCount == 0) 
-      {
-        oled_Print_Msg(0, "<DSMRloggerAPI>", 0);
-        oled_Print_Msg(1, "-------------------------",0);
-        oled_Print_Msg(2, "Raw Format",0);
-        sprintf(cMsg, "Raw Count %4d", showRawCount);
-        oled_Print_Msg(3, cMsg, 0);
-      }
-#endif
-
-      while(Serial.available() > 0) 
-      {   
-        char rIn = Serial.read();       
-        if (rIn == '!') 
-        {
-          showRawCount++;
-#if defined( HAS_OLED_SSD1306 ) || defined( HAS_OLED_SH1106 )
-          sprintf(cMsg, "Raw Count %4d", showRawCount);
-          oled_Print_Msg(3, cMsg, 0);
 #endif
-        }
-        TelnetStream.write((char)rIn);
-      }   // while Serial.available()
-      
-      if (showRawCount > 20) 
-      {
-        showRaw       = false;
-        showRawCount  = 0;
-      }
-  } 
+  blinkLEDms(1000);               // 'blink' the status led every x ms
+  yield();
 
-} // loop()
+}
 
+//===[ blink status led in ms ]===
+void blinkLEDms(uint32_t iDelay){
+  //blink the statusled, when time passed... #non-blocking blink
+  DECLARE_TIMER_MS(timerBlink, iDelay);
+  if (DUE(timerBlink))
+    blinkLEDnow();
+}
 
+//===[ blink status now ]===
+void blinkLEDnow(){
+    pinMode(LED_BUILTIN, OUTPUT);
+    digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
+}
 
+//===[ no-blocking delay with running background tasks in ms ]===
+void delayms(unsigned long delay_ms)
+{
+  DECLARE_TIMER_MS(timer, delay_ms);
+  while (!DUE(timer))
```

Comment:
Waarom denk je dat dit "non-blocking is"?

Reply:
Het is no-blocking in de zin dat de background taken actief aangeroepen worden gedurende de wachttijd. 

Het idee is dat je dus normaliter de loop aanroept, daar zitten de achtergrond taken. 

Vervolgens als je dan een “delay” wilt (wat bijna niet in de firmware zit trouwens) kan je deze functie gebruiken om je delay te doen. En dan heb je toch je achtergrond taken. 

Je hebt gelijk dat het niet non-blocking moet heten. Maar eigenlijk delay met background taken 😉